### PR TITLE
azure domain can be overridden with env variable

### LINKED ISSUE #2468: Lack of parameter "azure domain" when use azure blob storage as a repository
Output of `restic version`
--------------------------
restic 0.9.5 compiled with go1.11.5 on linux/amd64

Describe the issue
------------------
According to the manual here (https://restic.readthedocs.io/en/stable/030_preparing_a_new_repo.html#microsoft-azure-blob-storage),
we need to provide account name and access key when use azure block storage as a repository.
But that does not work if you azure domain is not azure global, like azure China and azure Germany. 
Error is returned as below:
> storage: service returned error: StatusCode=403, ErrorCode=AuthenticationFailed, ErrorMessage=Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.

The solution from Azure tech. support is, we must specify azure domain in this case as their tool Azure Storage Explore does.

### PR DESCRIPTION
**Pull request by arjunajesh**
What does this PR change? What problem does it solve?
-----------------------------------------------------
Restic backups on azure only worked for storages on the global domain `core.windows.net`. This meant that backups to other domains such as Azure China (`core.chinacloudapi.cn') were not supported. Restic now allows overriding the global domain using the environment variable `AZURE_DOMAIN'.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Closes #2468 

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [ ] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/backend/azure/config.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ type Config struct {
 	AccountName string
 	AccountSAS  options.SecretString
 	AccountKey  options.SecretString
+	Domain      string
```
Comment (MichaelEischer) [inline internal/backend/azure/config.go:19]: In the azure documentation, the parameter is called (storage) endpoint suffix.
https://learn.microsoft.com/en-us/azure/storage/common/storage-powershell-independent-clouds#get-endpoint-using-get-azenvironment

What about renaming it to `EndpointSuffix` and adjusting the environment variable name accordingly?

File: internal/backend/azure/azure.go [hunk 2]
```diff
@@ -53,7 +53,14 @@ func open(cfg Config, rt http.RoundTripper) (*Backend, error) {
...
 
-	url := fmt.Sprintf("https://%s.blob.core.windows.net/%s", cfg.AccountName, cfg.Container)
+	var domain string
+	if cfg.Domain != "" {
+		domain = cfg.Domain
+	} else {
+		domain = "core.windows.net"
+	}
+	url := fmt.Sprintf("https://%s.blob.%s/%s", cfg.AccountName, domain, cfg.Container)
+	fmt.Println(url)
```
Comment (MichaelEischer) [inline internal/backend/azure/azure.go:63]: Please remove the debug statement.

File: cmd/restic/cmd_backup.go [hunk 3]
```diff
@@ -541,7 +541,9 @@ func runBackup(ctx context.Context, opts BackupOptions, gopts GlobalOptions, ter
 	}
 
 	if !gopts.JSON {
+		fmt.Println("loading index now")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_backup.go:544]: Please revert the changes in this file.

File: changelog/unreleased/issue-2468 [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Issue: Compatibility with non global azure domain
+
+Restic backups on azure only worked for storages on the global domain `core.windows.net`. This meant that backups to other domains such as Azure China (`core.chinacloudapi.cn') were not supported. Restic now allows overriding the global domain using the environment variable `AZURE_DOMAIN'.
+
+https://github.com/restic/restic/issues/2468
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-2468:5]: Please add a link to this PR.

File: changelog/unreleased/issue-2468 [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Issue: Compatibility with non global azure domain
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-2468:1]: The prefix must be `Enhancement:`. I'd also reword the summary to `Enhancement: Add support for non-global Azure clouds`.

File: internal/backend/azure/config.go [hunk 6]
```diff
@@ -71,4 +72,7 @@ func (cfg *Config) ApplyEnvironment(prefix string) {
 	if cfg.AccountSAS.String() == "" {
 		cfg.AccountSAS = options.NewSecretString(os.Getenv(prefix + "AZURE_ACCOUNT_SAS"))
 	}
+	if cfg.Domain == "" {
+		cfg.Domain = os.Getenv("AZURE_DOMAIN")
```
Comment (MichaelEischer) [inline internal/backend/azure/config.go:76]: The environment variable should be included in the list in `doc/040_backup.rst`.

File: changelog/unreleased/issue-2468 [hunk 7]
```diff
@@ -0,0 +1,5 @@
...
+
+Restic backups on azure only worked for storages on the global domain `core.windows.net`. This meant that backups to other domains such as Azure China (`core.chinacloudapi.cn') were not supported. Restic now allows overriding the global domain using the environment variable `AZURE_DOMAIN'.
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-2468:3]: A similar description is also necessary in the Azure setup instructions in `doc/030_preparing_a_new_repo.rst`.

File: .DS_Store
Comment (MichaelEischer) [inline .DS_Store:1]: Please remove the .DS_Store files from the PR.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Thanks for working on this issue! The PR currently still needs a bit of polishing, see the comments below. The comments also explain where to add documentation. For testing it should be sufficient to manually verify that the parameter works as intended and does not change the current behavior is no other azure endpoint is specified.

**Comment by arjunajesh** [comment]
I have fixed everything you mentioned. Thanks for the help!

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!

I've squash the commits and added a few further tweaks.

**Pull request merged**